test/test-mongocrypt.c

```diff
@@ -225,6 +233,61 @@ bson_t *_mongocrypt_tester_bson_from_json(_mongocrypt_tester_t *tester, const ch
     return bson;
 }
 
+bson_t *tmp_bsonf(_mongocrypt_tester_t *tester, const char *fmt, ...) {
+    va_list arg;
+    va_start(arg, fmt);
+
+    char *dst = bson_malloc(strlen(fmt));
+    size_t dst_capacity = strlen(fmt);
+    size_t dst_len = 0;
+
+#define ENSURE_CAPACITY(len)                                                                                           \
+    if (1) {                                                                                                           \
+        if (len + dst_len >= dst_capacity) {                                                                           \
+            dst_capacity = len + dst_len;                                                                              \
+            dst_capacity *= 2;                                                                                         \
+            dst = bson_realloc(dst, dst_capacity);                                                                     \
+        }                                                                                                              \
+    } else                                                                                                             \
+        (void)0
+
+    for (const char *ptr = fmt; *ptr != '\0'; ptr++) {
+        if (0 == strncmp(ptr, "MC_STR", strlen("MC_STR"))) {
+            const char *src = va_arg(arg, const char *);
+            size_t src_len = strlen(src);
+            ENSURE_CAPACITY(src_len + 2);
+            *(dst + dst_len++) = '"';
+            strcpy(dst + dst_len, src);
+            dst_len += src_len;
+            *(dst + dst_len++) = '"';
+            ptr += strlen("MC_STR") - 1;
+            continue;
+        }
+
+        if (0 == strncmp(ptr, "MC_BSON", strlen("MC_BSON"))) {
+            const bson_t *src_bson = va_arg(arg, const bson_t *);
+            size_t src_len;
+            char *src = bson_as_canonical_extended_json(src_bson, &src_len);
+            ENSURE_CAPACITY(src_len);
+            strcpy(dst + dst_len, src);
+            dst_len += src_len;
+            bson_free(src);
+            ptr += strlen("MC_BSON") - 1;
+            continue;
+        }
+
+        ENSURE_CAPACITY(1);
+        *(dst + dst_len++) = *ptr;
```

Comment:
This loop could be improved by using `strstr` to search-and-directly-jump-to instances of `"MC_STR"` or `"MC_BSON"` (whichever comes first when present) instead of strcmp a per iteration (per character).